go/internal/tinder/driver_multi.go

```diff
@@ -48,126 +86,189 @@ func (md *MultiDriver) Advertise(ctx context.Context, ns string, opts ...p2p_dis
 	md.mapc[ns] = cf
 	md.muc.Unlock()
 
-	for _, driver := range md.drivers {
-		md.advertise(ctx, driver, ns, opts...)
+	ndrivers := len(md.drivers)
+	// We don't use a lock array because what we execute next is so cheap it's not worth the extra context switches that generates.
+	var wg sync.WaitGroup
+	wg.Add(ndrivers)
+	errReturn := make([]error, ndrivers)
+	durationReturn := make([]time.Duration, ndrivers)
+	for i := 0; i < ndrivers; i++ {
+		go func(j int) {
+			durationReturn[j], errReturn[j] = md.drivers[j].Advertise(ctx, ns, opts...)
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+
+	for i, v := range errReturn {
+		// Check if at least one was successful
+		if v == nil {
+			// One was, fetch the longer ttl
+			bestDuration := durationReturn[i]
+			for j := i + 1; j < ndrivers; j++ {
+				durj := durationReturn[j]
+				if durj > bestDuration {
+					bestDuration = durj
+				}
+			}
+			if options.Ttl != 0 && bestDuration > options.Ttl {
+				return options.Ttl, nil
+			}
+			return bestDuration, nil
+		}
 	}
 
-	return options.Ttl, nil
+	return 0, multierr.Combine(errReturn...)
 }
 
-func (md *MultiDriver) advertise(ctx context.Context, d Driver, ns string, opts ...p2p_discovery.Option) {
-	go func() {
-		for {
-			ttl, err := d.Advertise(ctx, ns, opts...)
-			if err != nil {
-				if ctx.Err() != nil {
-					return
-				}
+const maxLimit = 1000
 
-				md.logger.Warn("failed to advertise",
-					zap.String("driver", d.Name()),
-					zap.String("key", ns),
-					zap.Error(err),
-				)
+func (md *MultiDriver) FindPeers(ctx context.Context, ns string, opts ...p2p_discovery.Option) (<-chan p2p_peer.AddrInfo, error) {
+	var options p2p_discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
+	}
 
+	// Buffer to try to limit context switch while returning peers.
+	outPeers := make(chan p2p_peer.AddrInfo, 10)
```

Comment:
It's a round number. I've just avoided `-1` because I remember it as computationally intensive (might have changed since then) and `0` because that generates lots of contexts switch.
The buffer could be removed.

Reply:
I think 10 should be change to `limit`?

Reply:
> I think 10 should be change to `limit`?

Any value would work, I was just trying to not be wasteful on memory, so I've tested and I can't see any difference without creating any realistic setup.